### ISSUE TITLE
Core functions are not highlighted for ClojureScript

### ISSUE THREAD

**Issue by nikolavojicic** [clojure-emacs/cider#3385]
## Expected behavior

Core functions should be highlighted like `map` and `range` here. This is how it is for **Clojure**.

![Screenshot 2023-07-29 121408](https://github.com/clojure-emacs/cider/assets/10941031/3143bc86-acec-4f4a-a334-415ce5c9eb2e)

## Actual behavior

Core functions ( `map` and `range`) are not highlighted. This is how it is for **ClojureScript**.

![Screenshot 2023-07-29 121224](https://github.com/clojure-emacs/cider/assets/10941031/dfcac6f7-83c7-4eb9-b189-fadcc70d2af6)

## Steps to reproduce the problem

Content of `shadow-cljs.edn`:
```edn
{:source-paths ["src"]
 :dependencies [[binaryage/devtools "1.0.6"]
                [cider/cider-nrepl "0.32.0"]
                [net.cgrand/xforms "0.19.2"]
                [reagent "1.1.1"]
                [re-frame "1.2.0"]
                [cljs-http "0.1.46"]
                [day8.re-frame/http-fx "0.2.4"]
                [org.clojure/core.async "1.5.648"]]
 :builds       {:app {:target     :browser
                      :output-dir "public/js"
                      :asset-path "/js"
                      :modules    {:app {:entries [fisherman-fe.core]}}
                      :devtools   {:after-load fisherman-fe.core/init!}}}
 :dev-http     {3001 {:root    "public"
                      :handler user/app}}}
```

1. Run `cider-jack-in-cljs`
2. Choose `shadow`
3. Choose `:app`
4. Visit application in browser
5. Write the code above
6. Evaluate that code

## Environment & Version information

- shadow-cljs 2.20.13
- For CIDER and NREPL see screenshots above
- GNU Emacs 28.2 (build 2, x86_64-w64-mingw32) of 2022-09-13
- OpenJDK Runtime Environment Corretto-17.0.2.8.1 (build 17.0.2+8-LTS)
- Windows 10 (also in 11)

**Comment by bbatsov** [comment]
Why are you using `cider-nrepl` 0.32? I'm guessing that's some regression from the recent changes that @vemv has been doing. I guess he'll address this quickly once he sees the ticket.

**Comment by nikolavojicic** [comment]
@vemv advised me to try 0.32... Same behavior is with 0.30

![image](https://github.com/clojure-emacs/cider/assets/10941031/bed2466d-9689-4e24-bbb8-a85aa9ce6453)

**Comment by bbatsov** [comment]
Ah, okay. Well, I'll leave it to him to look into this. You might want to share here the part of your nREPL log related to `track-state`, as he'll likely need this to debug the problem.

**Comment by nikolavojicic** [comment]
Where can I find that log? I enabled nrepl message logging and this was printed: https://gist.github.com/nikolavojicic/fbd8eb603ca42ecc2d2acfce393d43cd

**Comment by vemv** [comment]
From my side, I have never gotten special syntax coloring for clojure.core defns, but that might be some oddity related to my setup.

Anyway, @bbatsov , do you recall what is _supposed_ to be happening, technically? Who flags clojure.core fns as such (cider-nrepl or cider), how? What font-lock face is applied to them?

**Comment by nikolavojicic** [comment]
Probably this setting does dynamic syntax coloring:
![image](https://github.com/clojure-emacs/cider/assets/10941031/19cd22fa-206d-46d8-8458-e3f30ea44c23)

**Comment by nikolavojicic** [comment]
Advanced syntax coloring occurs only after I run CIDER.

**Comment by nikolavojicic** [comment]
I've found this similar issue: https://github.com/clojure-emacs/cider/issues/1889#issuecomment-270370937

**Comment by vemv** [comment]
That's right! I had it set to `(cider-font-lock-dynamically '(macro deprecated))` long ago enough to completely forget about it :)

It's an accurate pointer, thanks. I'll begin investigating.

**Comment by vemv** [comment]
Alright, I had it easy to repro:

* `(cider-resolve-ns-symbols (cider-resolve-core-ns))` returns plenty of info when called from a jvm clj buffer
* `(cider-resolve-ns-symbols (cider-resolve-core-ns))` returns nil on cljs

This defun is buggy:

```lisp
(defun cider-resolve-core-ns ()
  "Return a dict of the core namespace for current connection.
This will be clojure.core or cljs.core depending on the return value of the
function `cider-repl-type'."
  (when-let* ((repl (cider-current-repl)))
    (with-current-buffer repl
      (cider-resolve--get-in (if (eq cider-repl-type 'cljs)
                                 "cljs.core"
                               "clojure.core")))))
```

...because `(eq cider-repl-type 'cljs)` is nil even on a .cljs buffer. A possible fix is `(or (eq cider-repl-type 'cljs) (eq major-mode 'clojurescript-mode))`.

Even after fixing that, `(cider-resolve-ns-symbols (cider-resolve-core-ns))` will be nil. That seems to be a deeper issue in the cider-nrepl middleware.

**Comment by vemv** [comment]
The second bug is this access https://github.com/clojure-emacs/cider-nrepl/blob/4b55dedf4cee10bda75ab2165580da72a617b721/src/cider/nrepl/middleware/track_state.clj#L222, that string should be a symbol.

With those two trivial bugs fixed, it should all work again 🍻

**Comment by vemv** [comment]
Please try again with cider-nrepl 0.34.0. Nothing else is needed

I believe this issue is solved by now, otherwise let me know!

**Comment by nikolavojicic** [comment]
Core functions are now highlighted but user-defined not.

<img width="943" alt="image" src="https://github.com/clojure-emacs/cider/assets/10941031/ead157c7-df22-4e4a-9099-4a3fefb42352">

**Comment by nikolavojicic** [comment]
I restarted cider, loaded that buffer and `increment` was highlighted. But new functions are not, until restarted. 

<img width="212" alt="image" src="https://github.com/clojure-emacs/cider/assets/10941031/97a9316e-a64a-4efd-881e-99b61a3842c4">

**Comment by vemv** [comment]
I cannot repro!

First I `(setq cider-font-lock-dynamically '(macro core function deprecated))`

Then I define a function `foo`, save the file, make sure that compilation succeeded.

Then, `function` highlighting will work immediately for all occurrences of `foo`.

The precise style that should be applied (per `M-x describe-char`) is `font-lock-function-name-face`. Does your theme customize it to a color of your liking?

If it persists, please create a new issue.

Cheers - V

**Comment by nikolavojicic** [comment]
The problem is that you font lock dynamically `'(macro core function deprecated)` while I do `'(macro core function var)`. With your setup I cannot repro either. So it seems that `var` makes the problem here.
